zebra-network/src/address_book.rs

```diff
@@ -214,129 +243,197 @@ impl AddressBook {
     }
 
     /// Returns true if the given [`SocketAddr`] has recently sent us a message.
-    pub fn recently_live_addr(&self, addr: &SocketAddr) -> bool {
+    pub fn recently_live_addr(&self, addr: SocketAddr) -> bool {
         let _guard = self.span.enter();
-        match self.by_addr.get(addr) {
+        match self.by_addr.get(&addr) {
             None => false,
-            // NeverAttempted, Failed, and AttemptPending peers should never be live
+            // Responded peers are the only peers that can be live
             Some(peer) => {
-                peer.last_connection_state == PeerAddrState::Responded
-                    && peer.get_last_seen() > AddressBook::liveness_cutoff_time()
+                peer.get_last_success().unwrap_or(chrono::MIN_DATETIME)
+                    > AddressBook::liveness_cutoff_time()
             }
         }
     }
 
-    /// Returns true if the given [`SocketAddr`] is pending a reconnection
-    /// attempt.
-    pub fn pending_reconnection_addr(&self, addr: &SocketAddr) -> bool {
+    /// Returns true if the given [`SocketAddr`] had a recent connection attempt.
+    pub fn recently_attempted_addr(&self, addr: SocketAddr) -> bool {
         let _guard = self.span.enter();
-        match self.by_addr.get(addr) {
+        match self.by_addr.get(&addr) {
             None => false,
-            Some(peer) => peer.last_connection_state == PeerAddrState::AttemptPending,
+            Some(peer) => {
+                peer.get_last_attempt().unwrap_or(chrono::MIN_DATETIME)
+                    > AddressBook::liveness_cutoff_time()
+            }
         }
     }
 
-    /// Returns true if the given [`SocketAddr`] might be connected to a node
-    /// feeding timestamps into this address book.
-    pub fn maybe_connected_addr(&self, addr: &SocketAddr) -> bool {
-        self.recently_live_addr(addr) || self.pending_reconnection_addr(addr)
+    /// Returns true if the given [`SocketAddr`] recently failed.
+    pub fn recently_failed_addr(&self, addr: SocketAddr) -> bool {
+        let _guard = self.span.enter();
+        match self.by_addr.get(&addr) {
+            None => false,
+            Some(peer) => {
+                peer.get_last_failed().unwrap_or(chrono::MIN_DATETIME)
+                    > AddressBook::liveness_cutoff_time()
+            }
+        }
     }
 
-    /// Return an iterator over all peers.
-    ///
-    /// Returns peers in reconnection attempt order, then recently live peers in
-    /// an arbitrary order.
-    pub fn peers(&'_ self) -> impl Iterator<Item = MetaAddr> + '_ {
+    /// Returns true if the given [`SocketAddr`] had recent attempts, successes,
+    /// or failures.
+    pub fn recently_used_addr(&self, addr: SocketAddr) -> bool {
+        self.recently_live_addr(addr)
+            || self.recently_attempted_addr(addr)
+            || self.recently_failed_addr(addr)
+    }
```

Comment:
Security: Limit reconnection rate to each individual peer address #1848

Part 1: A recently used peer is a peer that has had an attempt, success, or failure within the last 2 minutes

Reply:
This change is unlikely to be causing test failures, because:
* all seed peers get a connection attempt.

Unless we're somehow triggering a connection limit or firewall within our CI?
Maybe we should limit the initial connection rate.

Reply:
(If we were triggering CI issues, #2181 would be failing as well.)

Reply:
> Maybe we should limit the initial connection rate - this is a good idea anyway.

Done in #2181 for both the initial peers, and a fix to the `CandidateSet` peers.